Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialTypeBuilder.cpp

```diff
@@ -199,6 +268,269 @@ namespace AZ
+                                missingSomeFiles = true;
+                            }
+
+                            templateFilePath = resolvedFilePath;
+                        };
+
+                        resolveTemplateFilePathReference(materialPipelineFilePath, normalizedShaderTemplate.m_shader);
+                        resolveTemplateFilePathReference(materialPipelineFilePath, normalizedShaderTemplate.m_azsli);
+
+                        shaderTemplateReferences[normalizedShaderTemplate].push_back(materialPipelineName.c_str());
+                    }
+                }
+
+                if (missingSomeFiles)
+                {
+                    return;
+                }
+            }
+
+            // The new material type will no longer be abstract, we remove the reference to the partial
+            // material shader code and will replace it below with a concrete shader asset list.
+            const AZStd::string materialAzsliFilePath = AssetUtils::ResolvePathReference(request.m_sourceFile, materialType.m_materialShaderCode);
+            if (!AZ::IO::LocalFileIO::GetInstance()->Exists(materialAzsliFilePath.c_str()))
+            {
+                AZ_Error(MaterialTypeBuilderName, false, "File is missing: '%s'", materialType.m_materialShaderCode.c_str());
+                return;
+            }
+            materialType.m_materialShaderCode.clear();
+
+            if (!materialType.m_shaderCollection.empty())
+            {
+                // It technically is possible for the shader list to include shaders and have these combined with the auto-generated shaders below.
+                // However, there isn't a clear use case for this, so it is not allowed just in case it would introduce new edge cases. If a valid
+                // use case is presented, we could consider keeping the m_shaderCollection.
+                AZ_Warning(MaterialTypeBuilderName, false, "The shader list in the .materialtype file is not empty. This is unexpected and will be ignored.");
+                materialType.m_shaderCollection.clear();
+            }
+
+            // Generate the required shaders
+            for (const auto& [shaderTemplate, materialPipelineList] : shaderTemplateReferences)
+            {
+                AZStd::string materialPipelineName;
+
+                if (materialPipelineList.size() == 1)
+                {
+                    materialPipelineName = *materialPipelineList.begin();
+                }
+                else if(materialPipelineList.size() > 1)
+                {
+                    // Multiple material pipelines reference the same shader, so it should have a generic common name.
+                    materialPipelineName = PipelineNameForCommonShaders;
+                }
+                else
+                {
+                    AZ_Assert(false, "There should be at least one material pipeline referencing the shader");
```

Comment:
Question, are asset processor builders expected to handle errors differently through some error reporting API as opposed to asserting?

Reply:
AZ_Error is for problems in the data that a user can fix. AZ_Assert is for problems in the code that a programmer must fix. AZ_Assert is appropriate in this case, because it shouldn't be possible to get here since the loop that filled shaderTemplateReferences should always put at least one pipeline in the list. All add a code comment to clarify.